rosidl_typesupport_opensplice_c/resource/msg__type_support_c.cpp.template

```diff
@@ -0,0 +1,552 @@
+  DDS::DataReader * topic_reader = static_cast<DDS::DataReader*>(dds_data_reader);
+
+  @(__dds_msg_type_prefix)DataReader * data_reader =
+    @(__dds_msg_type_prefix)DataReader::_narrow(topic_reader);
+
+  @(__dds_msg_type_prefix)Seq dds_messages;
+  DDS::SampleInfoSeq sample_infos;
+  DDS::ReturnCode_t status = data_reader->take(
+    dds_messages,
+    sample_infos,
+    1,
+    DDS::ANY_SAMPLE_STATE,
+    DDS::ANY_VIEW_STATE,
+    DDS::ANY_INSTANCE_STATE);
+
+  const char * errs = nullptr;
+  bool ignore_sample = false;
+
+  switch (status) {
+    case DDS::RETCODE_ERROR:
+      errs = "@(__dds_msg_type_prefix)DataReader.take: "
+             "an internal error has occurred";
+      goto finally;
+    case DDS::RETCODE_ALREADY_DELETED:
+      errs = "@(__dds_msg_type_prefix)DataReader.take: "
+             "this @(__dds_msg_type_prefix)DataReader has already been deleted";
+      goto finally;
+    case DDS::RETCODE_OUT_OF_RESOURCES:
+      errs = "@(__dds_msg_type_prefix)DataReader.take: "
+             "out of resources";
+      goto finally;
+    case DDS::RETCODE_NOT_ENABLED:
+      errs = "@(__dds_msg_type_prefix)DataReader.take: "
+             "this @(__dds_msg_type_prefix)DataReader is not enabled";
+      goto finally;
+    case DDS::RETCODE_PRECONDITION_NOT_MET:
+      errs = "@(__dds_msg_type_prefix)DataReader.take: "
+             "a precondition is not met, one of: "
+             "max_samples > maximum and max_samples != LENGTH_UNLIMITED, or "
+             "the two sequences do not have matching parameters (length, maximum, release), or "
+             "maximum > 0 and release is false.";
+      goto finally;
+    case DDS::RETCODE_NO_DATA:
+      *taken = false;
+      errs = nullptr;
+      goto finally;
+    case DDS::RETCODE_OK:
+      break;
+    default:
+      errs = "@(__dds_msg_type_prefix)DataReader.take: unknown return code";
+      goto finally;
+  }
+
+  {
+    DDS::SampleInfo & sample_info = sample_infos[0];
+    if (!sample_info.valid_data) {
+      // skip sample without data
+      ignore_sample = true;
+    } else {
+      DDS::InstanceHandle_t sender_handle = sample_info.publication_handle;
+      auto sender_gid = u_instanceHandleToGID(sender_handle);
+      if (ignore_local_publications) {
+        // compare the system id from the sender and this receiver
+        // if they are equal the sample has been sent from this process and should be ignored
+        DDS::InstanceHandle_t receiver_handle = topic_reader->get_instance_handle();
+        auto receiver_gid = u_instanceHandleToGID(receiver_handle);
+        ignore_sample = sender_gid.systemId == receiver_gid.systemId;
+      }
+      // This is nullptr when being used with plain rmw_take, so check first.
+      if (sending_publication_handle) {
+        *static_cast<DDS::InstanceHandle_t *>(sending_publication_handle) = sender_handle;
+      }
+    }
+  }
+
+  if (!ignore_sample) {
+    errs = convert_dds_to_ros(&dds_messages[0], untyped_ros_message);
```

Comment:
I started to do that with the function pointers to the convert functions in the callback structure, but I decided to try and get something merged since we're so close to the alpha. I'm not particularly happy with it, but I don't think we have time to fix these things right now.